src/EditorFeatures/Core/Structure/StructureTag.cs

```diff
@@ -59,6 +61,25 @@ public StructureTag(AbstractStructureTaggerProvider tagProvider, BlockSpan block
         public bool IsDefaultCollapsed { get; }
         public bool IsImplementation { get; }
 
+        public override int GetHashCode()
+            => throw ExceptionUtilities.Unreachable();
+
+        public override bool Equals(object? obj)
+            => Equals(obj as StructureTag);
+
+        public bool Equals(StructureTag? other)
+        {
+            return other != null &&
+                this.GuideLineHorizontalAnchorPoint == other.GuideLineHorizontalAnchorPoint &&
+                this.Type == other.Type &&
+                this.IsCollapsible == other.IsCollapsible &&
+                this.IsDefaultCollapsed == other.IsDefaultCollapsed &&
+                this.IsImplementation == other.IsImplementation &&
+                _tagProvider.SpanEquals(this.Snapshot, this.OutliningSpan, other.Snapshot, other.OutliningSpan) &&
+                _tagProvider.SpanEquals(this.Snapshot, this.HeaderSpan, other.Snapshot, other.HeaderSpan) &&
+                _tagProvider.SpanEquals(this.Snapshot, this.GuideLineSpan, other.Snapshot, other.GuideLineSpan);
```

Comment:
What worries mere here specifically for this tagger is this is assuming the editor is mapping spans the same way we are. If they have any custom mapping logic (or just pick a different tracking mode), we could be in a case where we get out of sync with the editor where our interval tree says one thing and the editor's tracking is something else.

Reply:
so i'm not worried about that because the tagProvider *determiens* the tracking mode.  that's why i'm deferring to it to do the equality check.